lib/HLSL/DxilSimpleGVNHoist.cpp

```diff
@@ -585,3 +587,180 @@ FunctionPass *llvm::createDxilSimpleGVNHoistPass() {
+        Cond = Br->getCondition();
+      }
+      else if (SwitchInst *Sw = dyn_cast<SwitchInst>(TI)) {
+        Cond = Sw->getCondition();
+      }
+      // Don't know how to handle this. Don't assume anything.
+      else {
+        return false;
+      }
+
+      // Either add a record or compare it to the record
+      if (!Cmp) {
+        TerminatorRecord NewRec = {};
+        NewRec.ConditionVN = VT.lookupOrAdd(Cond);
+        NewRec.Opcode = TI->getOpcode();
+        TerminatorRecords.push_back(NewRec);
+      }
+      else {
+        if (TerminatorIndex >= TerminatorRecords.size())
+          return false;
+
+        TerminatorRecord Record = TerminatorRecords[TerminatorIndex];
+        if (Record.Opcode != TI->getOpcode() || Record.ConditionVN != VT.lookup(Cond, false))
+          return false;
+
+        TerminatorIndex++;
+      }
+    }
+    else {
+      // Either add a record or compare it to the record
+      if (!Cmp) {
+        uint32_t VN = VT.lookupOrAdd(I);
+        InstValueNumbers.push_back(VN);
+      }
+      else {
+        uint32_t VN = VT.lookupOrAdd(I);
+        if (InstIndex >= InstValueNumbers.size() || InstValueNumbers[InstIndex] != VN)
+          return false;
+        InstIndex++;
+      }
+    }
+
+    return true;
+  }
+
+  bool BuildOrCompare(BasicBlock *Begin, BasicBlock *End, ValueTable &VT, bool Compare) {
+    SmallVector<BasicBlock *, 10> Worklist;
+    DenseSet<BasicBlock *> Seen;
+    Worklist.push_back(Begin);
+
+    while (Worklist.size()) {
+      BasicBlock *BB = Worklist.pop_back_val();
+      if (BB == End)
+        continue;
+      Seen.insert(BB);
+
+      for (Instruction &I : *Begin) {
+        if (!HandleInst(VT, &I, Compare))
+          return false;
+      }
+
+      for (BasicBlock *Succ : successors(BB)) {
+        // If the successor was encountered before, there might be a loop in here somewhere.
+        // Don't risk it.
+        if (Seen.count(Succ)) {
+          return false;
+        }
+        Worklist.push_back(Succ);
+      }
+    }
+
+    if (!Compare) {
+      return true;
+    }
+    else {
+      return InstValueNumbers.size() == InstIndex && TerminatorRecords.size() == TerminatorIndex;
+    }
+  }
+
+};
+
+bool DxilSimpleGVNEliminateRegion::runOnFunction(Function &F) {
+  PostDominatorTree *PDT = &getAnalysis<PostDominatorTree>();
+
+  ValueTable VT;
+
+  bool bChanged = false;
+  for (BasicBlock &BB : F) {
+    TerminatorInst *TI = BB.getTerminator();
+    if (TI->getNumSuccessors() != 2) {
+      continue;
+    }
+
+    BasicBlock *Succ0 = TI->getSuccessor(0);
+    BasicBlock *Succ1 = TI->getSuccessor(1);
+    if (&BB == Succ0)
+      continue;
+    if (&BB == Succ1)
+      continue;
+    if (!HasOnePred(Succ0))
+      continue;
+    if (!HasOnePred(Succ1))
+      continue;
+
+    BasicBlock *End = PDT->findNearestCommonDominator(Succ0, Succ1);
+
+    ValueNumberComparator Comparison;
+    if (!Comparison.BuildOrCompare(Succ0, End, VT, /*compare*/false))
```

Comment:
A nicer api here would be to expose `Build()` and `Compare()` functions. Internally they could call `BuildOrCompare()`